standards/application/relay-reliability.md

```diff
@@ -0,0 +1,150 @@
+
+def checkOutgoingMessages():
+    for messageHash in outgoingMessageHashes:
+        message = database.getMessage(messageHash)
+        # only query store node for ongoing message, and posted more than 3 seconds ago
+        if message.status == 'ongoing' && time.now() - message.postTime > 3:
+            response = waku.store.queryMessage(messageHash)
+            if response.exists():
+                database.updateMessageStatus(messageHash, 'sent')
+                outgoingMessageHashes.remove(messageHash)
+            elif time.now() - message.postTime > 10:
+                # resend the message if it's not stored in store node after 10 seconds
+                waku.relay.post(message)
+```
+
+Function `checkOutgoingMessages` is called periodically, most likely every a few seconds. Message hashes can be queried in batch to reduce the number of requests to store nodes, the size in a batch shoud not exceed the max supported size by store node.
+
+The store node can be set and updated directly by application or selected from peers which are discovery by protocols like [discv5](https://github.com/vacp2p/rfc-index/blob/main/waku/standards/core/33/discv5.md) or [peer exchange](https://github.com/waku-org/specs/blob/master/standards/core/peer-exchange.md).
+
+The store node may only support specific pubsub topics, and the application should group message hashes by pubsub topic before sending the request.
+
+When persistent network issue happens, you may not want to resend the failed messages indefinitely, the application should have a mechanism to clean the cache with failed message hashes and trigger other retry logic after a few attempts.
+
+### Query with Topics and Time Range
+
+An application could use different pubsub topics and content topics, for example a community may have its own pubsub topic, and each channel may have its own content topic. To fetch all missing messages in a specific channel, the application can query the store node with the provided pubsub topic, content topic and time range.
+
+```python
+class FetchRecord:
+    pubsubTopic: str
+    lastFetch: int
+
+class QueryParams:
+    pubsubTopic: str
+    contentTopics: List[str]
+    fromTime: int
+    toTime: int
+
+def fetchMissingMessages(queryParams):
+    missingMessageHashes = []
+
+    # get missing message identifiers first in order to reduce the data transfer
+    response = waku.store.queryMessageHashes(queryParams)
+    for !response.isComplete():
+        # process each message in the response
+        response.messages().forEach(messageHash -> {
+            message = queryDbMessageByHash(messageHash)
+            if message.exists():
+                continue
+            }
+            missingMessageHashes.append(messageHash)
+        })
+
+        # process next page of the response
+        response.Next()
+    
+    # fetch missing messages with hashes in batch
+    response = waku.store.queryMessages(queryParams, missingMessageHashes)
```

Comment:
`queryParams` is not needed here, as `missingMessageHashes` will contain the primary keys for the table (it is also not supported). I also renamed the function to `queryMessagesByHash` to differentiate it from the other function as well as added a `peerID`

```suggestion
    response = waku.store.queryMessagesByHash(peerID, missingMessageHashes)
```